Add snap troubleshooting How-To (#943)
canonical/k8s-snap: docs/src/snap/howto/troubleshooting.md

# How to troubleshoot {{product}}

> This how-to will walk you through the steps to troubleshoot your Canonical Kubernetes cluster.

Reviewer: ```suggestion
This guide walks you through the steps to troubleshoot your Canonical Kubernetes cluster.
```

> ## Verify that the API server is healthy

Reviewer 2: I feel like we should start with the super high-level "cluster ready check" via `k8s status`

Verify that the API server is healthy and reachable by running the following command on a control-plane node:

```
sudo k8s kubectl get all
```

A failure could mean the API server on the particular node is unhealthy. The status of the API server service can be checked by running the following command:
```
sudo systemctl status snap.k8s.kube-apiserver
```

The logs of the API server service can be accessed by running the following command:
```
sudo journalctl -xe -u snap.k8s.kube-apiserver
```

If the API server is trying to be reached from a host that is not a control-plane node, a failure could mean:

> * API server is not reachable due to network issues or firewall limitations

Reviewer 2: nit: The API server ...
also below

> If the API server is trying to be reached from a host that is not a control-plane node, a failure could mean:
> * API server is not reachable due to network issues or firewall limitations
> * API server is failing on the control-plane node that's being reached
> * The control-plane node that's being reached is down

Reviewer: ```suggestion
If you are trying to reach the API server from a worker node, a failure could mean that:
* The API server is not reachable due to network issues or firewall limitations
* The API server is failing on the control-plane node that's being reached
* The control-plane node that's being reached is down
```

```{warning}

> The kubeconfig generated by the `k8s config` command uses the IP address of the control-plane node the command is executed on.

Reviewer: ```suggestion
When running `sudo k8s config` on a control-plane node you retrieve the kubeconfig file that uses this node's IP address.
```

```

> In this case try reaching the API server on a different control-plane node by updating the IP address that's used in the kubeconfig file.

Reviewer: ```suggestion
Try reaching the API server on a different control-plane node by updating the IP address that's used in the kubeconfig file.
```

 INFO:  Service k8s.containerd is running
 INFO:  Service k8s.kube-proxy is running
 INFO:  Service k8s.k8s-dqlite is running
 INFO:  Service k8s.k8sd is running
 INFO:  Service k8s.kube-apiserver is running
 INFO:  Service k8s.kube-controller-manager is running
 INFO:  Service k8s.kube-scheduler is running
 INFO:  Service k8s.kubelet is running
Collecting registry mirror logs
Collecting service arguments
 INFO:  Copy service args to the final report tarball
Collecting k8s cluster-info
 INFO:  Copy k8s cluster-info dump to the final report tarball
Collecting SBOM
 INFO:  Copy SBOM to the final report tarball
Collecting system information
 INFO:  Copy uname to the final report tarball
 INFO:  Copy snap diagnostics to the final report tarball
 INFO:  Copy k8s diagnostics to the final report tarball
Collecting networking information
 INFO:  Copy network diagnostics to the final report tarball
Building the report tarball
 SUCCESS:  Report tarball is at /root/inspection-report-20250109_132806.tar.gz
```

This confirms the services that are running, and the resulting report file can be viewed to get a detailed look at every aspect of the system.

## Common issues

> Please refer to [Troubleshooting Reference][snap-troubleshooting-reference] for common {{product}} issues.

Reviewer 2: This sounds like users are constantly facing issues with Canonical Kubernetes. I don't know how to phrase this better but @nhennigan or @louiseschmidtgen probably know how to let that sentence sound like we are not shipping a piece of garbage while still providing a reference to known issues.

Reviewer 3: I agree with putting this near the top and tell users to go there first. 

Maybe your issue has already been solved? Check out the [troubleshooting reference][snap-troubleshooting-reference] page to see a list of common issues and their solutions. Otherwise continue with this guide to help troubleshoot your {{product}}  cluster.

does that sound a bit better?

Reviewer 2: Sounds good

> ## Reporting a bug

Reviewer 2: I would put this right after the inspection report section as this is mostly the next step after creating a report. I'm not sure though where to put the reference to common issues. Maybe at the top of this page? "Please check if you face any of the [known issues] before proceeding with the section below".

Reviewer 3: +1

# How to troubleshoot {{product}}

Identifying issues in a Kubernetes cluster can be difficult, especially to new users. With {{product}} we aim to make deploying and managing your cluster as easy as possible. This how-to guide will walk you through the steps to troubleshoot your {{product}} cluster.

## Common issues

Maybe your issue has already been solved? Check out the [troubleshooting reference][snap-troubleshooting-reference] page to see a list of common issues and their solutions. Otherwise continue with this guide to help troubleshoot your {{product}} cluster.

> ## Verify that the cluster status is ready

Reviewer: ```suggestion
## Check the cluster status
```
Avoid having words in header & sentence duplicated

Verify that the cluster status is ready by running the following command:

```
sudo k8s status
```

> You should see output similar to the following:

Reviewer: ```suggestion
You should see a command output similar to the following:
```

```
cluster status:           ready
control plane nodes:      10.94.106.249:6400 (voter), 10.94.106.208:6400 (voter), 10.94.106.99:6400 (voter)
high availability:        yes
datastore:                k8s-dqlite
network:                  enabled
dns:                      enabled at 10.152.183.106
ingress:                  disabled
load-balancer:            disabled
local-storage:            enabled at /var/snap/k8s/common/rawfile-storage
gateway                   enabled
```

> ## Verify that the API server is healthy

Reviewer: ```suggestion
## Test the API server health
```

Verify that the API server is healthy and reachable by running the following command on a control-plane node:

```
sudo k8s kubectl get all
```

> This command lists resources that exist under the default namespace. You should see output similar to the following if the API server is healthy:

Reviewer: ```suggestion
This command lists resources that exist under the default namespace. You should see a command output similar to the following if the API server is healthy:
```

```
NAME                 TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)   AGE
service/kubernetes   ClusterIP   10.152.183.1   <none>        443/TCP   29m
```

A typical error message may look like this if the API server can not be reached:
```
The connection to the server 127.0.0.1:6443 was refused - did you specify the right host or port?
```

> A failure could mean the API server on the particular node is unhealthy. The status of the API server service can be checked by running the following command:

Reviewer: ```suggestion
A failure can mean that the API server on the particular node is unhealthy. Check the status of the API server service:
```

```
sudo systemctl status snap.k8s.kube-apiserver
```

> The logs of the API server service can be accessed by running the following command:

Reviewer: ```suggestion
Access the logs of the API server service by running the following command:
```
(Active voice)

```
sudo journalctl -xe -u snap.k8s.kube-apiserver
```

If you are trying to reach the API server from a host that is not a control-plane node, a failure could mean that:
* The API server is not reachable due to network issues or firewall limitations
* The API server is failing on the control-plane node that's being reached
* The control-plane node that's being reached is down

```{warning}
When running `sudo k8s config` on a control-plane node you retrieve the kubeconfig file that uses this node's IP address.
```

Try reaching the API server on a different control-plane node by updating the IP address that's used in the kubeconfig file.

> ## Verify that the cluster nodes are healthy

Reviewer: ```suggestion
## Check the cluster nodes' health
```

> Verify that the nodes in the cluster are healthy by running the following command and checking if all of the nodes have the `Ready` status:

Reviewer: ```suggestion
Confirm that the nodes in the cluster are healthy by looking for the `Ready` status:
```

```
sudo k8s kubectl get nodes
```

> 

Reviewer: I would also add an example output here

## Troubleshooting an unhealthy node

> There are certain services running on each node of a {{product}} cluster which are required. The required services depend on the type of node.

Reviewer: ```suggestion
Every healthy {{ product }} node has certain services up and running. The required services depend on the type of node.
```

> Services running on both types of nodes:

Reviewer: ```suggestion
Services running on both control plane and worker nodes:
```
Let's be specific when we can be.

* `k8sd`
* `kubelet`
* `containerd`
* `kube-proxy`

Services running only on control-plane nodes:
* `kube-apiserver`
* `kube-controller-manager`
* `kube-scheduler`
* `k8s-dqlite`

Services running only on worker nodes:
* `k8s-apiserver-proxy`

Check the status of these services on the failing node by running the following command:

> 

Reviewer: Should we add a check-all by running `sudo snap services k8s`?

Author: I don't consider this command too useful, systemd services usually restart and require closer inspection. I believe individual statuses provide more information.

```
sudo systemctl status snap.k8s.<service>
```

The logs of a failing service can be checked by running the following command:

```
sudo journalctl -xe -u snap.k8s.<service>
```

> If the issue indicates a problem with the configuration of the services on the node, it could be helpful to examine the arguments used to run these services.

Reviewer: ```suggestion
If the issue indicates a problem with the configuration of the services on the node, examine the arguments used to run these services.
```

> The arguments of a service on the failing node can be examined by reading the file located at `/var/snap/k8s/common/args/<service>`

Reviewer: ```suggestion
The arguments of a service on the failing node can be examined by reading the file located at `/var/snap/k8s/common/args/<service>`.
```
Punctuation

> ## Verify that the system pods are healthy

Reviewer: ```suggestion
## Investigating system pods' health
```

> Verify that the cluster's pods are healthy by running the following command and checking if all of the pods are `Running` and `Ready`:

Reviewer: ```suggestion
Check whether all of the cluster's pods are `Running` and `Ready`:
```
Short and sweet and avoid sentences starting in the same way.

> Verify that the pods that are a part of the cluster are healthy by running the following command and checking if all of the pods are `Running` and `Ready`:

Reviewer: ```suggestion
Verify that the cluster's pods are healthy by running the following command and checking if all of the pods are `Running` and `Ready`:
```

```
sudo k8s kubectl get pods -n kube-system
```

> The pods under the `kube-system` namespace belong to {{product}} features such as `network` that provide important functionality. Unhealthy pods could be related to configuration issues or nodes not meeting certain requirements.

Reviewer: ```suggestion
The pods in the `kube-system` namespace belong to {{product}}' features such as `network`. Unhealthy pods could be related to configuration issues or nodes not meeting certain requirements.
```

## Troubleshooting a failing pod

> Events on a failing pod can be checked by running the following command:

Reviewer: ```suggestion
Look at the events on a failing pod by running:
```
(Active voice)

```
sudo k8s kubectl describe pod <pod-name> -n <namespace>
```

> Logs on a failing pod can be checked by running the following command:

Reviewer: ```suggestion
Check the logs on a failing pod by running the following command:
```
(Active voice)

```
sudo k8s kubectl logs <pod-name> -n <namespace>
```

> 

Reviewer: Should we show how to look at specific containers in the pod- or does this go to far?

Author: I'm not sure where we should cut the line, this kind of turns into how to use kubectl in some ways. Should we assume a certain experience/expertise level here?

Reviewer: Fair enough- let's not dive too deep. We could add some external links to upstream troubleshooting at the end of this doc?

Author: Which section should we add the link to?

## Using the built-in inspection script

> {{product}} ships with a script to compile a complete report on {{product}} and the system which it is running on. This is an essential tool for bug reports and for investigating whether a system is (or isn’t) working and collecting all the relevant data in one place.

Reviewer: ```suggestion
{{product}} ships with a script to compile a complete report on {{product}} and its underlying system. This is an essential tool for bug reports and for investigating whether a system is (or isn’t) working.
```

> {{product}} ships with a script to compile a complete report on {{product}} and the system which it is running on. This is essential for bug reports, but is also a useful way of confirming the system is (or isn’t) working and collecting all the relevant data in one place.

Reviewer: ```suggestion
{{product}} ships with a script to compile a complete report on {{product}} and the system which it is running on. This is an essential tool for bug reports and for investigating whether a system is (or isn’t) working and collecting all the relevant data in one place.
```

> To run the inspection script, enter the command (admin privilege is required to collect all the data):

Reviewer: ```suggestion
Run the inspection script, by entering the command (admin privileges are required to collect all the data):
```

```
sudo /snap/k8s/current/k8s/scripts/inspect.sh
```

> You should see output similar to the following:

Reviewer: ```suggestion
The command output is similar to the following:
```

```
Collecting service information
Running inspection on a control-plane node
 INFO:  Service k8s.containerd is running
 INFO:  Service k8s.kube-proxy is running
 INFO:  Service k8s.k8s-dqlite is running
 INFO:  Service k8s.k8sd is running
 INFO:  Service k8s.kube-apiserver is running
 INFO:  Service k8s.kube-controller-manager is running
 INFO:  Service k8s.kube-scheduler is running
 INFO:  Service k8s.kubelet is running
Collecting registry mirror logs
Collecting service arguments
 INFO:  Copy service args to the final report tarball
Collecting k8s cluster-info
 INFO:  Copy k8s cluster-info dump to the final report tarball
Collecting SBOM
 INFO:  Copy SBOM to the final report tarball
Collecting system information
 INFO:  Copy uname to the final report tarball
 INFO:  Copy snap diagnostics to the final report tarball
 INFO:  Copy k8s diagnostics to the final report tarball
Collecting networking information
 INFO:  Copy network diagnostics to the final report tarball
Building the report tarball
 SUCCESS:  Report tarball is at /root/inspection-report-20250109_132806.tar.gz
```

> This script confirms that the services are running, and the resulting report file can be viewed to get a detailed look at every aspect of the system.

Reviewer: ```suggestion
Use the report to ensure that all necessary services are running and dive into every aspect of the system.
```

> This confirms the services that are running, and the resulting report file can be viewed to get a detailed look at every aspect of the system.

Reviewer: ```suggestion
This script confirms that the services are running, and the resulting report file can be viewed to get a detailed look at every aspect of the system.
```

## Reporting a bug

> If you cannot solve your issue and believe the fault may lie in {{product}}, please [file an issue on the project repository][].

Reviewer: ```suggestion
If you cannot solve your issue and believe that the fault may lie in {{product}}, please [file an issue on the project repository][].
```

> To help us deal effectively with issues, it is very useful to include the report obtained from the inspect script, as well as any additional logs, and a summary of the issue.

Reviewer: ```suggestion
Help us deal effectively with issues by including the report obtained from the inspect script, any additional logs, and a summary of the issue.
```
